src/pwm/hrtim.rs

```diff
@@ -0,0 +1,772 @@
+                    cortex_m::interrupt::free(|_| {
+                        let common = unsafe { &*HRTIM_COMMON::ptr() };
+                        common.oenr.modify(|_r, w| { w.$tXYoen().clear_bit() });
+                    });
+                }
+            }
+        )+
+    }
+}
+
+macro_rules! hrtim_out_common {
+    ($TIMX:ident, $set_event:ident, $register:ident, $action:ident) => {{
+        let tim = unsafe { &*$TIMX::ptr() };
+
+        match $set_event {
+            EventSource::Cr1 => tim.$register.modify(|_r, w| w.cmp1().$action()),
+            EventSource::Cr2 => tim.$register.modify(|_r, w| w.cmp2().$action()),
+            EventSource::Cr3 => tim.$register.modify(|_r, w| w.cmp3().$action()),
+            EventSource::Cr4 => tim.$register.modify(|_r, w| w.cmp4().$action()),
+            EventSource::Period => tim.$register.modify(|_r, w| w.per().$action()),
+        }
+    }};
+}
+
+macro_rules! hrtim_out {
+    ($($TIMX:ident: $out_type:ident: $tXYoen:ident, $setXYr:ident, $rstXYr:ident,)+) => {$(
+        impl HrOutput for $out_type<$TIMX> {
+            fn enable(&mut self) {
+                cortex_m::interrupt::free(|_| {
+                    let common = unsafe { &*HRTIM_COMMON::ptr() };
+                    common.oenr.modify(|_r, w| { w.$tXYoen().set_bit() });
+                });
+            }
+
+            fn disable(&mut self) {
+                cortex_m::interrupt::free(|_| {
+                    let common = unsafe { &*HRTIM_COMMON::ptr() };
+                    common.oenr.modify(|_r, w| { w.$tXYoen().clear_bit() });
+                });
+            }
+
+            fn enable_set_event(&mut self, set_event: EventSource) {
+                hrtim_out_common!($TIMX, set_event, $setXYr, set_bit)
+            }
+            fn disable_set_event(&mut self, set_event: EventSource) {
+                hrtim_out_common!($TIMX, set_event, $setXYr, clear_bit)
+            }
+
+            fn enable_rst_event(&mut self, reset_event: EventSource) {
+                hrtim_out_common!($TIMX, reset_event, $rstXYr, set_bit)
+            }
+            fn disable_rst_event(&mut self, reset_event: EventSource) {
+                hrtim_out_common!($TIMX, reset_event, $rstXYr, clear_bit)
+            }
+        }
+    )+};
+}
+
+hrtim_out! {
+    HRTIM_TIMA: HrOut1: ta1oen, seta1r, rsta1r,
+    HRTIM_TIMA: HrOut2: ta2oen, seta2r, rsta2r,
+
+    HRTIM_TIMB: HrOut1: tb1oen, setb1r, rstb1r,
+    HRTIM_TIMB: HrOut2: tb2oen, setb2r, rstb2r,
+
+    HRTIM_TIMC: HrOut1: tc1oen, setc1r, rstc1r,
+    HRTIM_TIMC: HrOut2: tc2oen, setc2r, rstc2r,
+
+    HRTIM_TIMD: HrOut1: td1oen, setd1r, rstd1r,
+    HRTIM_TIMD: HrOut2: td2oen, setd2r, rstd2r,
+
+    HRTIM_TIME: HrOut1: te1oen, sete1r, rste1r,
+    HRTIM_TIME: HrOut2: te2oen, sete2r, rste2r,
+
+    // TODO: Somehow, there is no rstf1r
+    //HRTIM_TIMF: HrOut1: tf1oen, setf1r, rstf1r,
+
+    // TODO: Somehow, there is no tf2oen
+    //HRTIM_TIMF: HrOut2: tf2oen, setf2r, rstf2r,
```

Comment:
A bit weird, [the reference](https://www.st.com/resource/en/reference_manual/rm0440-stm32g4-series-advanced-armbased-32bit-mcus-stmicroelectronics.pdf#page=1031) says there should be a `tf2oen`... Is this a bug in the crate `stm32g4`? I believe the `HRTIM_TIMF` timer might be unique to `stm32g474`/`stm32g484` thus missing in `stm32f334` and `stm32h7`.